src/bootstrap/src/utils/build_stamp.rs

```diff
@@ -0,0 +1,194 @@
+//! Module for managing build stamp files.
+//!
+//! Contains the core implementation of how bootstrap utilizes stamp files on build processes.
+
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+
+use sha2::digest::Digest;
+
+use crate::core::builder::Builder;
+use crate::core::config::TargetSelection;
+use crate::utils::helpers::{hex_encode, mtime};
+use crate::{Compiler, Mode, helpers, t};
+
+#[cfg(test)]
+mod tests;
+
+/// Manages a stamp file to track build state. The file is created in the given
+/// directory and can have custom content and name.
+#[derive(Clone)]
+pub struct BuildStamp {
+    path: PathBuf,
+    pub(crate) stamp: String,
+}
+
+impl AsRef<Path> for BuildStamp {
+    fn as_ref(&self) -> &Path {
+        &self.path
+    }
+}
```

Comment:
Nit: I find this `AsRef<Path>` somewhat strange, maybe just have a dedicated method?

```rs
impl BuildStamp {
    fn path(&self) -> &Path { &self.path }
}
```

Reply:
(Duplicated review comment)